### PR TITLE
MatchedClient Tests, Rename, and Documentation Updates

### LINKED ISSUE #120: Revise "Clients" documentation
I propose that clients.rst, and other portions of the docs, be revised with the following in mind:

- The basic control of Agents using a MatchedClient should be covered, or at least referenced, in the User Guide.
- Terminology around "Client" is confusing... is a Client an object used to talk to Agents, or is it a script/program/daemon used to "orchestrate multi-agent interactions"?  Certainly "MatchedClient" should refer to that object/class, and not also to some application that makes use of that class.  This terminology originates with the original OCS concept of a "Control Client"...  perhaps we should amend that terminology to something more like "Control Program" (which could be a script or something more persistent). 
- I think we should move MatchedClient to the top, and explicitly encourage the use of MatchedClients for interactive use (jupyter, ipython).  "Basic Clients" should be mentioned as an option for longer-lived control processes (small p), such as servos or monitoring applications.
- Should we rename "Basic Clients" -> "Control Programs using Twisted"?

### LINKED ISSUE #222: Rename MatchedClients and organize Client development docs
Stemming from discussion in #13, we plan to rename `MatchedClient` to be, simply `OCSClient`. This will take precedence in the documentation for writing clients. What we are currently calling the "Basic Client" will be updated as "Twisted Client", since that structure is required if you want to write an asynchronous client, and will be secondary to the OCSClient in the docs, as it is less commonly used.

`MatchedClient` will be left in as a simple wrapper to `OCSClient` to not break backwards compatibility, but will be marked as deprecated, and possibly removed in a later version.

More examples for interacting with clients, particularly in passing parameters to them, will also be added.

### PR DESCRIPTION
**Pull request by BrianJKoopman**
## Description

This PR contains three primary updates:

- Development of unit tests for `ocs.matched_client` in it's original form.
  - This was done to ensure when I renamed `MatchedClient`, I didn't break anything.
  - Once the rename was successful, and tests completed, I moved the tests to appropriately test the new `OCSClient` object instead (in 512585b). Note that tests were successful before this migration on 1488871.
  - While writing these tests I made some small changes to functions in here, including marking functions as private when used only within `ocs.matched_client`, and removing unused arguments.
- Renaming of `MatchedClient` to `OCSClient`, as decided in #222.
  - For ease of reading for the actual rename see 3b4d65f.
- Rewrite of Developer Guide page on Clients.
  - This now prioritizes the OCSClient object, attempts to distinguish the concept of "Client" and "Control Program", and is now linked in the User Guide's Quickstart.

For the review, it'd be helpful to get a read of the new Clients documentation page. Any comments on the renaming and/or testing are certainly welcome as well.

Note: I'm PRing this onto the in progress testing branches, as despite it being mostly separate, it uses some of the tooling built in those PRs. If those cannot be reviewed before this, I can rework this dependence.

## Motivation and Context

- Fixes #222 
- Fixes #120

## How Has This Been Tested?

The tests have been run, updated to use `OCSClient` everywhere.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] Unless I am preparing a release, I have opened this PR onto the `develop` branch.

### CODE REVIEW

File: docs/developer/clients.rst [hunk 1]
```diff
@@ -2,27 +2,321 @@
...
+crossbar server and makes the "remote procedure call" to an Agent. Multiple
+Clients can be used to orchestrate an observatory within a Control Program
+("program" or "script").
+
+While there are other options for how a Client could be implemented, the focus
+of this documentation will be on using the OCSClient object within a script.
+Other options are discussed below.
+
+.. _ocs_client:
+
```
Comment (mhasself) [inline docs/developer/clients.rst:19]: This (somewhere in Clients or OCSClient) could/should mention ocs-client-cli.  And ocs-client-cli docs could link back here.

**Review by mhasself — APPROVED** [review]
Brave rewrite! 

There remain some stray uses of MatchedClient in the source tree:
```
ocs/client_cli.py
example/docs/agent_template.rst
example/miniobs/run_acq.py
```

**Review by jlashner — APPROVED** [review]
This looks fantastic. Reading through the docs really makes me appreciate how far things have come since the early days!

File: ocs/matched_client.py [hunk 2]
```diff
@@ -1,169 +1,12 @@
...
-        instance_id (str): Instance id for agent to run
-
-    """
-
+@deprecation.deprecated(
+    deprecated_in='v0.9.0',
+    details="Renamed to OCSClient"
+)
+class MatchedClient(OCSClient):
     def __init__(self, instance_id, **kwargs):
```
[the remark this answers is not in the document]
Comment (BrianJKoopman) [reply ocs/matched_client.py:11]: Ah, yup, that works.

File: tests/test_matched_client.py [hunk 3]
```diff
@@ -1,28 +1,15 @@
-import os
-from unittest.mock import MagicMock, patch
-from ocs.matched_client import MatchedClient
-
-mocked_client = MagicMock()
-mock_from_yaml = MagicMock()
+from unittest.mock import patch
 
+from ocs.matched_client import MatchedClient
```
[the remark this answers is not in the document]
Comment (BrianJKoopman) [reply tests/test_matched_client.py:3]: I think until `matched_client.py` is removed in some later version, yes...

File: docs/developer/clients.rst [hunk 4]
```diff
@@ -2,27 +2,321 @@
...
+This starts the Task and then immediately waits for it to complete (assuming
+the task starts successfully), equivalent to::
+
+    response = client.delay_task.start(delay=1)
+    if response[0] == ocs.OK:
+        client.delay_task.wait()
+
+Direct calls to a Process behave a bit differently, acting as an alias to
+"status", making these two calls are identical::
```
[the remark this answers is not in the document]
Comment (BrianJKoopman) [reply docs/developer/clients.rst:113]: Dropped the 'are' instead.

File: docs/developer/clients.rst [hunk 5]
```diff
@@ -2,27 +2,321 @@
...
+OCSWeb Client
+`````````````
+
+A Client can be written in JavaScript. This is what is done in OCS Web. For
+more details about how to implement a JavaScript Client, see
+:ref:`creating_web_panel`.
+
+
+Control Programs using Twisted
+``````````````````````````````
```
[the remark this answers is not in the document]
Comment (BrianJKoopman) [reply docs/developer/clients.rst:304]: Great point, added a sentence on this to the opening paragraph.

**Comment by BrianJKoopman** [comment]
> Brave rewrite!
> 
> There remain some stray uses of MatchedClient in the source tree:
> 
> ```
> ocs/client_cli.py
> example/docs/agent_template.rst
> example/miniobs/run_acq.py
> ```

Thanks, I swapped those instances (and one more that I found in the tests) with `OCSClient`.

Thanks both for the review, will merge shortly.

**Pull request merged**